src/conversation_engine/engine.py

```diff
@@ -0,0 +1,409 @@
+                                                feedback_cb=self._feedback_cb)
+            rospy.loginfo("Task sent: {}".format(self._state.current_semantics))
+
+            self._state.wait_for_robot()
+        else:
+            rospy.loginfo("Could not parse '{}'".format(text))
+            self._log_invalid_command(text)
+
+            if 'sandwich' in text:
+                result_sentence = "Try 'sudo {}'.".format(text)
+            else:
+                result_sentence = random.choice(["You're not making sense.",
+                                                 "Don't give me this shit.",
+                                                 "Tell me something useful.",
+                                                 "This is useless input. Thanks, but no thanks.",
+                                                 "Make sense to me, fool!",
+                                                 "Talk to the gripper, the PC is too good for you.",
+                                                 "Something went terribly wrong.",
+                                                 "Would your mother in law understand?",
+                                                 "Try 'sudo {}'.".format(text)])
+            self._robot_to_user_pub.publish(result_sentence)
+            self._start_new_conversation()
+
+    def _handle_additional_info(self, text):
+        """Parse text into additional info according to grammar & target received from action_server result"""
+        rospy.loginfo("_handle_additional_info('{}')".format(text))
+
+        words = text.strip().split(" ")
+        additional_semantics = self._parser.parse(self._state.target, words, debug=True)
+
+        rospy.loginfo("additional_semantics: {}".format(additional_semantics))
+        sem_str = json.dumps(additional_semantics)
+        sem_dict = yaml.load(sem_str)
+        rospy.logdebug("parsed: {}".format(sem_dict))
+
+        if additional_semantics:
+            try:
+                rospy.loginfo("Additional_semantics: {}".format(additional_semantics))
+                self._state.update_semantics(sem_dict, self._state.missing_field)
+                self._robot_to_user_pub.publish(random.choice(["OK, I can work with that",
+                                                               "Allright, thanks!"]))
+
+                self._action_client.send_async_task(str(self._state.current_semantics),
+                                                    done_cb=self._done_cb,
+                                                    feedback_cb=self._feedback_cb)
+                rospy.loginfo("Updated task sent: {}".format(self._state.current_semantics))
+
+                self._state.wait_for_robot()
+            except Exception as e:
+                rospy.logerr("Could not update semantics: {}".format(e))
+                self._robot_to_user_pub.publish("Something went terribly wrong, can we try a new command?")
```

Comment:
'A new command' isn't very natural. Something like: 'What would you like me to do?'?